File: Sorting Plateau/Binary Search/Q2. Binary Search/Q2_BinarySearch.py
class Solution(object):                  # Define Solution class

    def search(self, nums, target):      # Function to search target

        """
        :type nums: List[int]            # nums is a list of integers
        :type target: int                # target is an integer
        :rtype: int                      # return type is integer
        """

        lo = 0                           # Start index
        hi = len(nums) - 1               # End index

        while lo <= hi:                  # Loop while range is valid

            mid = (lo + hi) // 2         # Find middle index

            if nums[mid] == target:      # If target found
                return mid               # Return index

            if target > nums[mid]:       # If target is bigger
                lo = mid + 1             # Search right side
            else:                        # If target is smaller
                hi = mid - 1             # Search left side

        return -1                        # Return -1 if not found

File: Sorting Plateau/Binary Search/Q2. Binary Search/Q2_BinarySearch.java
class Solution {                          // Define the Solution class

    public int search(int[] nums, int target) { // Function to search target

        int low = 0;                      // Start index of array
        int high = nums.length - 1;       // End index of array

        while (low <= high) {             // Loop while range is valid

            int mid = (low + high) / 2;   // Find middle index

            if (nums[mid] == target) {    // If target is found
                return mid;               // Return index
            }
            else if (nums[mid] > target) { // If middle value is greater
                high = mid - 1;           // Search left side
            }
            else {                        // If middle value is smaller
                low = mid + 1;            // Search right side
            }
        }

        return -1;                        // Return -1 if target not found
    }
}
